content/docs/iac/guides/building-extending/providers/provider-architecture.md

```diff
@@ -0,0 +1,97 @@
+
+At this layer, you implement the provider interface directly. You write your [schema JSON](/docs/iac/guides/building-extending/packages/schema/) by hand, giving you complete control over resource definitions, property types, and how data structures map between Pulumi and your provider. This explicit control over the schema is particularly valuable when you need precise type mappings or want to define resources that don't fit standard patterns.
+
+Choose this layer when you need a provider in a language without a higher-level SDK, when you want full control over your schema and data structure mappings, when you're building something unusual that doesn't fit SDK patterns, or when you prefer explicit protocol handling over framework abstractions.
+
+For implementation guides, see [Direct implementation in Python](/docs/iac/guides/building-extending/providers/implementers/python/), [Direct implementation in Go](/docs/iac/guides/building-extending/providers/implementers/go/), or [Direct implementation in TypeScript](/docs/iac/guides/building-extending/providers/implementers/typescript/).
+
+### Layer 3: Higher-level SDKs
+
+SDKs build on the language bindings to provide a more ergonomic development experience. Currently, the [Pulumi Go Provider SDK](/docs/iac/guides/building-extending/providers/sdks/pulumi-go-provider-sdk/) is available for Go.
+
+The Go SDK provides schema inference from Go structs and tags, default implementations for most methods, middleware support for cross-cutting concerns, a testing framework for provider validation, and automatic multi-language SDK generation.
+
+Choose this layer when you're writing a provider in Go and want the fastest path to a working provider. It's ideal when you prefer convention over configuration and want automatic schema generation.
+
+## Choosing your layer
+
+| Consideration | Layer 2 (Direct) | Layer 3 (SDK) |
+|---------------|------------------|---------------|
+| **Language** | Any with gRPC support | Go only (currently) |
+| **Schema** | Hand-written JSON | Inferred from code |
+| **Boilerplate** | More | Less |
+| **Control** | Full | Framework-guided |
+| **Learning curve** | Protocol knowledge required | Go SDK patterns |
+
+### Choosing a language
+
+Many teams prefer to write providers in their standard language, giving them access to familiar tools and libraries. However, Go has a significant advantage: providers written in Go compile to standalone binaries with no runtime dependencies. Users of your provider can consume it from any Pulumi language without installing Go.
+
+Providers written in Python or TypeScript require their respective runtimes to be installed, which adds friction for users. That said, if your team is more productive in Python or TypeScript, the familiar ecosystem and libraries may outweigh the runtime dependency—especially for internal providers where you control the deployment environment.
```

Comment:
Not sure it's worth mentioning, but it's possible to create a Python or TypeScript provider that's wrapped in a native binary. Not a very common thing to do, but it's possible.

Reply:
What do recommend for doing that?